src/job_metadata.c

```diff
@@ -650,3 +692,245 @@ PgCronHasBeenLoaded(void)
+	Oid argTypes[MAX_NUMBER_SPI_EXEC_ARGS];
+	Datum argValues[MAX_NUMBER_SPI_EXEC_ARGS];
+	int i;
+
+	SetCurrentStatementStartTimestamp();
+	StartTransactionCommand();
+	PushActiveSnapshot(GetTransactionSnapshot());
+
+	if (!PgCronHasBeenLoaded() || RecoveryInProgress())
+	{
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+		ereport(LOG,(errmsg("pg_cron not loaded/present or recovery in progress")));
+		return;
+	}
+
+	initStringInfo(&querybuf);
+	i = 0;
+
+	/* Open SPI context. */
+	if (SPI_connect() != SPI_OK_CONNECT)
+		elog(ERROR, "SPI_connect failed");
+
+
+	appendStringInfo(&querybuf,
+		"update %s.%s set", CRON_SCHEMA_NAME, JOB_RUN_DETAILS_TABLE_NAME);
+
+
+	/* add the fields to be updated */
+	if (job_pid != NULL) {
+		argTypes[i] = INT4OID;
+		argValues[i] = Int32GetDatum(*job_pid);
+		i++;
+		appendStringInfo(&querybuf, " job_pid = $%d,", i);
+	}
+
+	if (status != NULL)
+	{
+		argTypes[i] = TEXTOID;
+		argValues[i] = CStringGetTextDatum(status);
+		i++;
+
+		appendStringInfo(&querybuf, " status = $%d,", i);
+	}
+
+        if (return_message != NULL)
+	{
+		argTypes[i] = TEXTOID;
+		argValues[i] = CStringGetTextDatum(return_message);
+		i++;
+
+		appendStringInfo(&querybuf, " return_message = $%d,", i);
+	}
+
+        if (start_time != NULL)
+	{
+		argTypes[i] = TIMESTAMPTZOID;
+		argValues[i] = TimestampTzGetDatum(*start_time);
+		i++;
+
+		appendStringInfo(&querybuf, " start_time = $%d,", i);
+	}
+
+        if (end_time != NULL)
+	{
+		argTypes[i] = TIMESTAMPTZOID;
+		argValues[i] = TimestampTzGetDatum(*end_time);
+		i++;
+
+		appendStringInfo(&querybuf, " end_time = $%d,", i);
+	}
+
+	argTypes[i] = INT8OID;
+	argValues[i] = Int64GetDatum(runId);
+	i++;
+
+	/* remove the last comma */
+	querybuf.len--;
+	querybuf.data[querybuf.len] = '\0';
+
+	/* and add the where clause */
+	appendStringInfo(&querybuf, " where runid = $%d", i);
+
+	pgstat_report_activity(STATE_RUNNING, querybuf.data);
+
+	if(SPI_execute_with_args(querybuf.data,
+		i, argTypes, argValues, NULL, false, 1) != SPI_OK_UPDATE)
+		elog(ERROR, "SPI_exec failed: %s", querybuf.data);
+
+	pfree(querybuf.data);
+
+	SPI_finish();
+	PopActiveSnapshot();
+	CommitTransactionCommand();
+	pgstat_report_activity(STATE_IDLE, NULL);
+}
+
+void
+MarkPendingRunsAsFailed(void)
+{
+	StringInfoData querybuf;
+
+	SetCurrentStatementStartTimestamp();
+	StartTransactionCommand();
+	PushActiveSnapshot(GetTransactionSnapshot());
+
+	if (!PgCronHasBeenLoaded() || RecoveryInProgress())
+	{
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+		return;
+	}
+
+	initStringInfo(&querybuf);
+
+	/* Open SPI context. */
+	if (SPI_connect() != SPI_OK_CONNECT)
+		elog(ERROR, "SPI_connect failed");
+
+
+	appendStringInfo(&querybuf,
+		"update %s.%s set status = '%s', return_message = 'server restarted' where status in ('%s','%s')"
+		, CRON_SCHEMA_NAME, JOB_RUN_DETAILS_TABLE_NAME, GetCronStatus(CRON_STATUS_FAILED), GetCronStatus(CRON_STATUS_STARTING), GetCronStatus(CRON_STATUS_RUNNING));
```

Comment:
it's usually nicer to use quote_qualified_identifier to generate qualified table names.